add button component, scaffold demo and build system

potch opened the pull request:
* Added button component
* Pulled in other photon repos for icons and color
* Created specimen page
* Wrote script to perform vendoring so dependencies can be moved to devDependencies

potch commented:
What the specimen looks like:

<img width="953" alt="screenshot 2018-08-17 14 09 47" src="https://user-images.githubusercontent.com/253470/44288991-37e55200-a227-11e8-9339-cf3b83746c33.png">

bwinton commented:
Huh... I thought we were going to provide new [WebComponents](https://developer.mozilla.org/en-US/docs/Web/Web_Components), not style default HTML tags. What do other people think? Benefits/disadvantages?

potch commented:
@bwinton the BEM components can be easily wrapped into Web Components, but not vice versa. Pure CSS/HTML will allow these to be adapted to a number of component formats, and I'd expect that we can check Web Component versions of these into this repo as well.

bgrins commented:
> Huh... I thought we were going to provide new WebComponents, not style default HTML tags. What do other people think? Benefits/disadvantages?

I can more easily see pulling the photon css / markup in to existing Custom Elements in Firefox chrome than pulling in an entire photon Custom Element. Since we (currently) rely on xul-specific stuff and (may always) rely on chrome-specific stuff for the components used in chrome it'll be tricky to vendor an entire Custom Element in and use it directly, if that makes sense. Although I guess a photon Custom Element could be directly used as part of a chrome Custom Element - I'll have to think about that more.

IMO it's worth reviewing the HTML/CSS as-is to make sure it's visually correct - we can see what migrating it to a Custom Element would look like separately. And we can hold off on publishing / advertising the repo until we have a consensus on that.

bwinton commented on button/button.css:
```diff
@@ -0,0 +1,116 @@
+/* Primary style */
+
+button.button--primary,
+a.button.button--primary {
```
Do we need both `.button` and `.button--primary`? I would figure that the second kinda implies the first...

bwinton replied:
(This might be a philosophical objection I have to BEM, but I also feel like the Firefox front-end team will have a similar objection to adding a ton of classes where one would do. 🙂)

potch replied:
The notion (from BEM) is that a "primary" button is a *kind* of button, vs a whole new thing. so you say "hey this is a button", and also "it's the primary kind". I'm open to changing it in the name of ergonomics ("They're more guidelines than rules..."), but it does in practice reduce the risk of namespace conflicts.

johngruen replied:
I'm a big fan of BEM for long term legibility. It's really not terribly longwinded at the end of the day, and it allows for strongly scoped styles.

bwinton replied:
Okay, if everyone else likes it, we can keep the extra classes. 🤷‍♂️🙂

potch commented:
@aminalhazwani let me know if you need any context around this PR or have any questions.

aminalhazwani commented on button/button.css:
```diff
@@ -0,0 +1,116 @@
+button:focus,
+a.button:focus {
+  box-shadow:
+    0 0 0 1px #0a84ff inset,
```
Why didn't we opt for `var(--blue-50)` color variable here? Same below.

potch replied:
wasn't sure those matched from the photon design docs, will definitely use that.

aminalhazwani commented on demo.css:
```diff
@@ -0,0 +1,52 @@
+
+body {
+  background-color: var(--grey-10);
+  font-family: 'Fira Sans', sans-serif;
```
We are starting to slowly deprecate Fira Sans in favor of system fonts, we could do something like this:

```
font-family: -apple-system, BlinkMacSystemFont,
             'helvetica neue', helvetica,
             ubuntu,
             roboto, noto,
             'segoe ui', arial,
             sans-serif;
```

or like this (for firefox-only products):

```
font: message-box;
```

potch replied:
Is this preference true for general web stuff as well as in-product?

aminalhazwani replied:
Yes, that is correct, in-product we use system fonts. New Tab has been using system fonts since the AS beginnings and Firefox Accounts has also recently moved to system fonts, and they are planning to do the same for AMO.

aminalhazwani commented on button/button.css:
```diff
@@ -0,0 +1,116 @@
+  font-weight: 400;
+  height: 32px;
+  line-height: 32px;
+  min-width: 132px;
```
This can be optional as there might be cases where there isn't enough space for respecting the `min-width`.

aminalhazwani commented on button/button.css:
```diff
@@ -0,0 +1,116 @@
+
+button:hover,
+a.button:hover {
+  background-color: var(--grey-90-a20);
```
`.button--ghost:hover` should be `var(--grey-90-a10)` instead of `var(--grey-90-a20)`

potch replied:
the design docs do say that, but looking at real examples, they all seem to be a20

aminalhazwani replied:
Can you point me to an example? Because looking at the toolbar buttons they set the hover state to `hsla(240, 5%, 5%, 0.1)` which is `var(--grey-90-a10)`.

<img width="77" alt="screen shot 2018-09-19 at 9 36 32 am" src="https://user-images.githubusercontent.com/1721875/45738019-9f5c3c00-bbef-11e8-9288-122f4ac9df94.png">

potch replied:
Given that I can't even find the supposed example I was working from 😬, I'll change this to --grey-90-a10

potch replied:
This should be good now!

johngruen commented on README.md:
```diff
@@ -1,2 +1,14 @@
+
+## Conventions
+
+Components are styled using the BEM (Block-Element-Modifier) convention:
```
Add a link to the docs: http://getbem.com/

potch commented:
Working on addressing feedback today.

potch commented:
@aminalhazwani for now I'm going to split the demo page out to a separate folder within the repo, and we can split it fully into its own place in the future.

potch commented:
I've added a follow-up commit that addresses comments here and elsewhere:
* restructured project, moving demos to a specific `demo` folder
* renamed `button` to `photon-buttons` to match future package names
* put a readme in `photon-buttons`
* moved button content layout to inline-flex vs line-height
* used color variables where available

potch commented on photon-buttons/README.md:
````diff
@@ -0,0 +1,61 @@
+### Default
+
+```html
+  &lt;a href="#" class="button">Anchor&lt;/a>
````
Did this in the commit below!

aminalhazwani commented:
Cool! I think we can merge @bwinton!

On the topic of demos, over the weekend I found this neat docs generator from the Paciello Group https://thepaciellogroup.github.io/cupper/patterns/coding/writing-inline-demos/, worth looking into it!

aminalhazwani commented:
Also this is interesting https://thepaciellogroup.github.io/cupper/patterns/coding/code-blocks/#annotated-code 🙂
